crates/blockchain/blockchain.rs

```diff
@@ -161,6 +143,120 @@ pub fn validate_block(
+    let head_header_res = store.get_block_by_hash(head_hash).map_err(wrap)?;
+
+    // Check that we already have all the needed blocks stored and that we have the ancestors
+    // if we have the descendants, as we are working on the assumption that we only add block
+    // if they are connected to the canonical chain.
+    let (finalized, safe, head) = match (finalized_header_res, safe_header_res, head_header_res) {
+        (None, Some(_), _) => return Err(InvalidForkChoice::ElementNotFound),
+        (_, None, Some(_)) => return Err(InvalidForkChoice::ElementNotFound),
+        (Some(f), Some(s), Some(h)) => (f.header, s.header, h.header),
+        _ => return Err(InvalidForkChoice::Syncing),
+    };
+
+    // Check that we are not being pushed pre-merge
+    total_difficulty_check(&head_hash, &head, store)?;
+
+    // Check that the headers are in the correct order.
+    if finalized.number > safe.number || safe.number > head.number {
+        return Err(InvalidForkChoice::Unordered);
+    }
+
+    // If the head block is already in our canonical chain, the beacon client is
+    // probably resyncing. Ignore the update.
+    if is_canonical(store, head.number, head_hash).map_err(wrap)? {
+        return Ok(());
+    }
+
+    // Find out if blocks are correctly connected.
+    let Some(new_canonical_blocks) = find_link_with_canonical_chain(store, &head).map_err(wrap)?
+    else {
+        return Err(InvalidForkChoice::Disconnected(
+            error::ForkChoiceElement::Head,
+            error::ForkChoiceElement::Safe,
+        ));
+    };
+
+    let link_block_number = match new_canonical_blocks.last() {
+        Some((number, _)) => *number,
+        None => head.number,
+    };
+
+    // Check that finalized and safe blocks are either in the new canonical blocks, or already
+    // but prior to the canonical link to the new head. This is a relatively quick way of making
+    // sure that head, safe and finalized are connected.
+
+    if !(is_canonical(store, finalized.number, finalized_hash).map_err(wrap)?
+        && finalized.number <= link_block_number
+        || new_canonical_blocks.contains(&(finalized.number, finalized_hash))
+        || (finalized.number == head.number && finalized_hash == head_hash))
+    {
+        return Err(InvalidForkChoice::Disconnected(
+            error::ForkChoiceElement::Head,
+            error::ForkChoiceElement::Finalized,
+        ));
+    };
+
+    if !(is_canonical(store, safe.number, safe_hash).map_err(wrap)?
+        && safe.number <= link_block_number
+        || new_canonical_blocks.contains(&(safe.number, safe_hash))
+        || (safe.number == head.number && safe_hash == head_hash))
+    {
+        return Err(InvalidForkChoice::Disconnected(
+            error::ForkChoiceElement::Head,
+            error::ForkChoiceElement::Safe,
+        ));
+    };
+
+    // Finished all validations.
+    for (number, hash) in new_canonical_blocks {
```

Comment:
what if you have this situation:

canonical chain:

1 - 2 - 3 - 4

And you get a new FCU with a new head 3'

Then the canonical chain should end up being:
1 - 2 - 3'

and not: 
1 - 2 - 3' - 4